File: src/main/java/AoC2024_14.java
import static com.github.pareronia.aoc.StringOps.splitLines;
import static com.github.pareronia.aoc.StringOps.splitOnce;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;

import com.github.pareronia.aoc.IntegerSequence.Range;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.geometry.Vector;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_14
        extends SolutionBase<List<AoC2024_14.Robot>, Integer, Integer> {
    
    private static final int W = 101;
    private static final int H = 103;
    
    private AoC2024_14(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_14 create() {
        return new AoC2024_14(false);
    }
    
    public static AoC2024_14 createDebug() {
        return new AoC2024_14(true);
    }
    
    @Override
    protected List<Robot> parseInput(final List<String> inputs) {
        return inputs.stream().map(Robot::fromInput).toList();
    }
    
    private int getSafetyFactor(
            final List<Robot> robots,
            final int w,
            final int h,
            final int rounds
    ) {
        final int midW = w / 2;
        final int midH = h / 2;
        final int[] q = {0, 0, 0, 0};
        robots.stream()
            .map(r -> r.position.translate(r.velocity, rounds))
            .forEach(p -> {
                final int px = Math.floorMod(p.getX(), w);
                final int py = Math.floorMod(p.getY(), h);
                if (px < midW) {
                    if (py < midH) {
                        q[0] += 1;
                    } else if (midH < py) {
                        q[1] += 1;
                    }
                } else if (midW < px) {
                    if (py < midH) {
                        q[2] += 1;
                    } else if (midH < py) {
                        q[3] += 1;
                    }
                }
        });
        return Arrays.stream(q).reduce(1, (a, b) -> a * b);
    }
    
    private int solve1(final List<Robot> robots, final int w, final int h) {
        return getSafetyFactor(robots, w, h, 100);
    }
    
    @Override
    public Integer solvePart1(final List<Robot> robots) {
        return solve1(robots, W, H);
    }
    
    @Override
    public Integer solvePart2(final List<Robot> robots) {
        int ans = 0;
        int best = Integer.MAX_VALUE;
        int round = 1;
        final List<Integer> sfs = new ArrayList<>();
        while (round <= W + H) {
            final int safetyFactor = getSafetyFactor(robots, W, H, round);
            sfs.add(safetyFactor);
            if (safetyFactor < best) {
                best = safetyFactor;
                ans = round;
            }
            round++;
        }
        final List<Integer> mins = new ArrayList<>();
        Range.between(2, sfs.size() - 3).forEach(i -> {
            final int avg = Set.of(-2, -1, 1, 2).stream()
                    .mapToInt(j -> sfs.get(i + j)).sum() / 4;
            if (sfs.get(i) < avg * 9 / 10) {
                mins.add(i + 1);
            }
        });
        final int period = mins.get(2) - mins.get(0);
        round = mins.get(2) + period;
        while (round <= W * H) {
            final int safetyFactor = getSafetyFactor(robots, W, H, round);
            if (safetyFactor < best) {
                best = safetyFactor;
                ans = round;
            }
            round += period;
        }
        return ans;
    }
    
    @Override
    public void samples() {
        final AoC2024_14 test = AoC2024_14.createDebug();
        final List<Robot> input = test.parseInput(splitLines(TEST));
        assert test.solve1(input, 11, 7) == 12;
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_14.create().run();
    }

    private static final String TEST = """
            p=0,4 v=3,-3
            p=6,3 v=-1,-3
            p=10,3 v=-1,2
            p=2,0 v=2,-1
            p=0,0 v=1,3
            p=3,0 v=-2,-2
            p=7,6 v=-1,-3
            p=3,0 v=-1,-2
            p=9,3 v=2,3
            p=7,3 v=-1,2
            p=2,4 v=2,-3
            p=9,5 v=-3,-3
            """;

    record Robot(Position position, Vector velocity) {

        public static Robot fromInput(final String line) {
            final StringSplit split = splitOnce(line, " ");
            final StringSplit p = splitOnce(
                    splitOnce(split.left(), "=").right(),
                    ",");
            final StringSplit v = splitOnce(
                    splitOnce(split.right(), "=").right(),
                    ",");
            return new Robot(
                Position.of(
                        Integer.parseInt(p.left()),
                        Integer.parseInt(p.right())),
                Vector.of(
                        Integer.parseInt(v.left()),
                        Integer.parseInt(v.right()))
            );
        }
    }
}

File: src/main/python/AoC2024_14.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 14
#

import math
import sys

from aoc.common import InputData
from aoc.common import SolutionBase

Position = tuple[int, int]
Vector = tuple[int, int]
Input = list[tuple[Position, Vector]]
Output1 = int
Output2 = int

W, H = 101, 103

TEST = """\
p=0,4 v=3,-3
p=6,3 v=-1,-3
p=10,3 v=-1,2
p=2,0 v=2,-1
p=0,0 v=1,3
p=3,0 v=-2,-2
p=7,6 v=-1,-3
p=3,0 v=-1,-2
p=9,3 v=2,3
p=7,3 v=-1,2
p=2,4 v=2,-3
p=9,5 v=-3,-3
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        robots = list[tuple[Position, Vector]]()
        for line in input_data:
            p, v = line.split()
            px, py = map(int, p.split("=")[1].split(","))
            vx, vy = map(int, v.split("=")[1].split(","))
            robots.append(((px, py), (vx, vy)))
        return robots

    def safety_factor(
        self,
        robots: list[tuple[Position, Vector]],
        w: int,
        h: int,
        rounds: int,
    ) -> int:
        mid_w, mid_h = w // 2, h // 2
        q1, q2, q3, q4 = 0, 0, 0, 0
        for pos, vec in robots:
            px, py = pos[0] + rounds * vec[0], pos[1] + rounds * vec[1]
            px, py = px % w, py % h
            if px < mid_w:
                if py < mid_h:
                    q1 += 1
                elif mid_h < py:
                    q2 += 1
            elif mid_w < px:
                if py < mid_h:
                    q3 += 1
                elif mid_h < py:
                    q4 += 1
        return math.prod((q1, q2, q3, q4))

    def solve_1(
        self, robots: list[tuple[Position, Vector]], w: int, h: int
    ) -> int:
        return self.safety_factor(robots, w, h, 100)

    def part_1(self, robots: Input) -> Output1:
        return self.solve_1(robots, W, H)

    def part_2(self, robots: Input) -> Output2:
        def print_robots(
            robots: list[tuple[Position, Vector]], w: int, h: int, round: int
        ) -> None:
            if not __debug__:
                return
            new_robots = [
                ((pos[0] + round * vec[0]) % w, (pos[1] + round * vec[1]) % h)
                for pos, vec in robots
            ]
            lines = [
                "".join("*" if (x, y) in new_robots else "." for x in range(w))
                for y in range(h)
            ]
            for line in lines:
                print(line)

        ans, best, round = 0, sys.maxsize, 1
        sfs = list[int]()
        while round < W + H:
            safety_factor = self.safety_factor(robots, W, H, round)
            sfs.append(safety_factor)
            if safety_factor < best:
                best = safety_factor
                ans = round
            round += 1
        mins = list[int]()
        for i in range(2, len(sfs) - 2):
            avg = sum(sfs[i + j] for j in (-2, -1, 1, 2)) // 4
            if sfs[i] < avg * 9 // 10:
                mins.append(i + 1)
        period = mins[2] - mins[0]
        round = mins[2] + period
        while round <= W * H:
            safety_factor = self.safety_factor(robots, W, H, round)
            if safety_factor < best:
                best = safety_factor
                ans = round
            round += period
        print_robots(robots, W, H, ans)
        return ans

    def samples(self) -> None:
        assert self.solve_1(self.parse_input(TEST.splitlines()), 11, 7) == 12


solution = Solution(2024, 14)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
